File: io.cucumber/java8/src/test/java/cucumber/runtime/java8/test/LambdaStepdefs.java
package cucumber.runtime.java8.test;

import cucumber.api.DataTable;
import cucumber.api.Scenario;
import cucumber.api.java8.En;

import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotSame;

public class LambdaStepdefs implements En {
    private static LambdaStepdefs lastInstance;

    public LambdaStepdefs() {
        Before((Scenario scenario) -> {
            assertNotSame(this, lastInstance);
            lastInstance = this;
        });

        Given("^this data table:$", (DataTable peopleTable) -> {
            List<Person> people = peopleTable.asList(Person.class);
            assertEquals("Hellesøy", people.get(0).last);
        });

        Integer alreadyHadThisManyCukes = 1;

        Given("^I have (\\d+) cukes in my belly$", (Long n) -> {
            assertEquals((Integer) 1, alreadyHadThisManyCukes);
            assertEquals((Long) 42L, n);
        });

        String localState = "hello";
        Then("^I really have (\\d+) cukes in my belly", (Integer i) -> {
            assertEquals((Integer) 42, i);
            assertEquals("hello", localState);
        });

        int localInt = 1;
        Given("^A statement with a simple match$", () -> {
            assertEquals(2, localInt+1);
        });

        Given("^I will give you (\\d+) and ([\\d\\.]+) and (\\w+) and (\\d+)$", (Integer a, Float b, String c,
                                                                                 Integer d)
                -> {
            assertEquals((Integer) 1, a);
            assertEquals((Float) 2.2f, b);
            assertEquals("three", c);
            assertEquals((Integer) 4, d);
        });
    }

    public static class Person {
        String first;
        String last;
    }
}

File: kotlin-java8/src/test/kotlin/cucumber/runtime/kotlin/test/LambdaStepdefs.kt
package cucumber.runtime.kotlin.test;

import cucumber.api.Scenario
import io.cucumber.datatable.DataTable
import cucumber.api.java8.En
import org.junit.Assert.*

var lastInstance : LambdaStepdefs? = null

class LambdaStepdefs : En {

    init {
        Before { scenario: Scenario ->
            assertNotSame(this, lastInstance)
            lastInstance = this
        }

        BeforeStep { scenario: Scenario ->
            assertSame(this, lastInstance)
            lastInstance = this
        }

        AfterStep { scenario: Scenario ->
            assertSame(this, lastInstance)
            lastInstance = this
        }

        After { scenario: Scenario ->
            assertSame(this, lastInstance)
            lastInstance = this
        }

        Given("this data table:") { peopleTable: DataTable ->
            val people : List<Person> = peopleTable.asList(Person::class.java)
            assertEquals("Aslak", people[0].first)
            assertEquals("Hellesøy", people[0].last)
        }

        val alreadyHadThisManyCukes = 1
        Given("I have {long} cukes in my belly") { n: Long ->
            assertEquals(1, alreadyHadThisManyCukes)
            assertEquals(42L, n)
        }

        val localState = "hello"
        Then("I really have {int} cukes in my belly") { i: Int ->
            assertEquals(42, i)
            assertEquals("hello", localState)
        }

        Given("A statement with a body expression$") { assertTrue(true) }

        Given("A statement with a simple match$", { -> assertTrue(true) })

        val localInt = 1
        Given("A statement with a scoped argument$", { assertEquals(2, localInt + 1) })

        Given("I will give you {int} and {float} and {word} and {int}") { a: Int, b: Float, c: String, d: Int ->
            assertEquals(1, a)
            assertEquals(2.2f, b)
            assertEquals("three", c)
            assertEquals(4, d)
        }
    }

    class Person {
        var first: String? = null
        var last: String? = null
    }
}
